File: services/federation-product-exchange-service/src/models/space.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::models::decode_json;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NexusSpace {
    pub id: uuid::Uuid,
    pub slug: String,
    pub display_name: String,
    pub description: String,
    pub space_kind: String,
    pub owner_peer_id: Option<uuid::Uuid>,
    pub region: String,
    pub member_peer_ids: Vec<uuid::Uuid>,
    pub governance_tags: Vec<String>,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, sqlx::FromRow)]
pub struct SpaceRow {
    pub id: uuid::Uuid,
    pub slug: String,
    pub display_name: String,
    pub description: String,
    pub space_kind: String,
    pub owner_peer_id: Option<uuid::Uuid>,
    pub region: String,
    pub member_peer_ids: Value,
    pub governance_tags: Value,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

impl TryFrom<SpaceRow> for NexusSpace {
    type Error = String;

    fn try_from(row: SpaceRow) -> Result<Self, Self::Error> {
        Ok(Self {
            id: row.id,
            slug: row.slug,
            display_name: row.display_name,
            description: row.description,
            space_kind: row.space_kind,
            owner_peer_id: row.owner_peer_id,
            region: row.region,
            member_peer_ids: decode_json(row.member_peer_ids, "member_peer_ids")?,
            governance_tags: decode_json(row.governance_tags, "governance_tags")?,
            status: row.status,
            created_at: row.created_at,
            updated_at: row.updated_at,
        })
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct CreateSpaceRequest {
    pub slug: String,
    pub display_name: String,
    pub description: String,
    pub space_kind: String,
    pub owner_peer_id: Option<uuid::Uuid>,
    pub region: String,
    #[serde(default)]
    pub member_peer_ids: Vec<uuid::Uuid>,
    #[serde(default)]
    pub governance_tags: Vec<String>,
    pub status: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct UpdateSpaceRequest {
    pub display_name: Option<String>,
    pub description: Option<String>,
    pub owner_peer_id: Option<uuid::Uuid>,
    pub region: Option<String>,
    pub member_peer_ids: Option<Vec<uuid::Uuid>>,
    pub governance_tags: Option<Vec<String>>,
    pub status: Option<String>,
}

File: services/tenancy-organizations-service/internal/models/space.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// NexusSpace mirrors `models::space::NexusSpace`.
type NexusSpace struct {
	ID             uuid.UUID   `json:"id"`
	Slug           string      `json:"slug"`
	DisplayName    string      `json:"display_name"`
	Description    string      `json:"description"`
	SpaceKind      string      `json:"space_kind"`
	OwnerPeerID    *uuid.UUID  `json:"owner_peer_id"`
	Region         string      `json:"region"`
	MemberPeerIDs  []uuid.UUID `json:"member_peer_ids"`
	GovernanceTags []string    `json:"governance_tags"`
	Status         string      `json:"status"`
	CreatedAt      time.Time   `json:"created_at"`
	UpdatedAt      time.Time   `json:"updated_at"`
}

// CreateSpaceRequest is the body of POST /spaces.
//
// Rust uses `#[serde(default)]` on `member_peer_ids` and `governance_tags`,
// so a missing field decodes as the empty slice — Go's nil-slice unmarshal
// already matches that.
type CreateSpaceRequest struct {
	Slug           string      `json:"slug"`
	DisplayName    string      `json:"display_name"`
	Description    string      `json:"description"`
	SpaceKind      string      `json:"space_kind"`
	OwnerPeerID    *uuid.UUID  `json:"owner_peer_id,omitempty"`
	Region         string      `json:"region"`
	MemberPeerIDs  []uuid.UUID `json:"member_peer_ids"`
	GovernanceTags []string    `json:"governance_tags"`
	Status         string      `json:"status"`
}

// UpdateSpaceRequest is the body of PATCH /spaces/:id.
type UpdateSpaceRequest struct {
	DisplayName    *string      `json:"display_name,omitempty"`
	Description    *string      `json:"description,omitempty"`
	OwnerPeerID    *uuid.UUID   `json:"owner_peer_id,omitempty"`
	Region         *string      `json:"region,omitempty"`
	MemberPeerIDs  *[]uuid.UUID `json:"member_peer_ids,omitempty"`
	GovernanceTags *[]string    `json:"governance_tags,omitempty"`
	Status         *string      `json:"status,omitempty"`
}
